### PR TITLE
JDK-8316778: test hprof lib: invalid array element type from JavaValueArray.elementSize

### PR DESCRIPTION
**Pull request by alexmenkov**
The change fixes 2 issues in hprof test library.
The issue were discovered during test development (logging values of dumped heap objects).
- JavaValueArray.elementSize cannot determine size of the array elements and throws RuntimeException: invalid array element type
- JavaObject.toString() method for dumped String objects tries to construct string value of the object from "value" field assuming "value" field is array of char. Actually it's byte array, so toString return array values like "{65, 66, 67}" (with fixed JavaValueArray.elementSize)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316778](https://bugs.openjdk.org/browse/JDK-8316778): test hprof lib: invalid array element type from JavaValueArray.elementSize (**Bug** - P4)

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**) ⚠️ Review applies to [2adcc81c](https://git.openjdk.org/jdk/pull/15953/files/2adcc81c12cf62aa6fce8bc7abf2a151bd594af9)
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15953/head:pull/15953` \
`$ git checkout pull/15953`

Update a local copy of the PR: \
`$ git checkout pull/15953` \
`$ git pull https://git.openjdk.org/jdk.git pull/15953/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15953`

View PR using the GUI difftool: \
`$ git pr show -t 15953`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15953.diff">https://git.openjdk.org/jdk/pull/15953.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15953#issuecomment-1738334708)

### CODE REVIEW

**Review by plummercj — APPROVED** [review]
Just a minor nit. Otherwise looks good.

File: test/lib-test/jdk/test/lib/hprof/HprofTest.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+import jdk.test.lib.hprof.model.JavaClass;
+import jdk.test.lib.hprof.model.JavaHeapObject;
+import jdk.test.lib.hprof.model.JavaObject;
+import jdk.test.lib.hprof.model.JavaValueArray;
+import jdk.test.lib.hprof.model.JavaThing;
+import jdk.test.lib.hprof.model.Snapshot;
+import jdk.test.lib.hprof.parser.Reader;
+
+/**
+ * @test
```
[the remark this answers is not in the document]
Comment (alexmenkov) [reply test/lib-test/jdk/test/lib/hprof/HprofTest.java:43]: Done.

File: test/lib-test/jdk/test/lib/hprof/HprofTest.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+ * @bug 8316778
+ * @library /test/lib
+ * @run main/othervm HprofTest
```
Comment (lmesnik) [inline test/lib-test/jdk/test/lib/hprof/HprofTest.java:46]: Why othervm is needed for the test? Wouldn't just main enough?

**Review by lmesnik — CHANGES_REQUESTED** [review]

**Review by lmesnik — APPROVED** [review]

File: test/lib-test/jdk/test/lib/hprof/HprofTest.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        }
+        int instanceCount = jClass.getInstancesCount(false);
+        if (instanceCount < 1) {
+            throw new RuntimeException("Not instances of '" + className + "' found");
+        }
+        log("Found " + instanceCount + " instance(s).");
+        return jClass.getInstances(false);
+    }
+
+    private static void log(Object s) {
```
Comment (sspitsyn) [inline test/lib-test/jdk/test/lib/hprof/HprofTest.java:168]: Nit: Why is the arg of `Object` type? Also, all calls to `log()` always pass `String` type.

**Review by sspitsyn — APPROVED** [review]
Looks good.
Thanks,
Serguei

File: test/lib-test/jdk/test/lib/hprof/HprofTest.java (hunk 3 above)
Comment (alexmenkov) [reply test/lib-test/jdk/test/lib/hprof/HprofTest.java:168]: Sometimes it it's useful to have ability to log arbitrary objects (or primitive values which are auto-boxed) without explicit conversion to String

**Pull request closed without merging**